### PR TITLE
[OSRM] New Binary

### PR DESCRIPTION
**Pull request by jeremiahpslewis**
# Release plan:
- [x] Once oneTBB PR https://github.com/JuliaPackaging/Yggdrasil/pull/6115 is merged, reactivate windows
- [ ] Get a new OSRM release tagged https://github.com/Project-OSRM/osrm-backend/issues/6516
- [ ] Merge this PR and ship binary for linux
- [ ] Once mac builds are debugged, then add and ship them

### CODE REVIEW

File: O/OSRM/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+cmake .. \
+    -DBZIP2_INCLUDE_DIR=${includedir} \
+    -DBZIP2_LIBRARIES=${libdir}/libbz2.${dlext} \
+    -DLUA_INCLUDE_DIR=${includedir} \
+    -DLUA_LIBRARIES=${libdir}/liblua.${dlext} \
+    -DZLIB_INCLUDE_DIR=${includedir} \
+    -DZLIB_LIBRARY=${libdir}/libz.${dlext} \
+    -Wno-dev
+cmake --build .
+cmake --build . --target install
```
Comment (giordano) [inline O/OSRM/build_tarballs.jl:31]: Parallel build would probably make this less horribly slow.

File: O/OSRM/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+# The products that we will ensure are always built
+products = Product[
+    LibraryProduct("libosrm", :libosrm)
+    ExecutableProduct("osrm-routed", :osrm_routed)
+    
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="Bzip2_jll", uuid="6e34b625-4abd-537c-b88f-471c36dfa7a0"))
```
Comment (giordano) [inline O/OSRM/build_tarballs.jl:47]: #3024

File: O/OSRM/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    Dependency(PackageSpec(name="XML2_jll", uuid="02c8fc9c-b97f-50b9-bbe4-9be30ff0a78a"))
+    Dependency(PackageSpec(name="boost_jll", uuid="28df3c45-c428-5900-9ff8-a3135698ca75"))
```
Comment (giordano) [inline O/OSRM/build_tarballs.jl:49]: #3024

File: O/OSRM/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    HostBuildDependency(PackageSpec(name="Lua_jll", uuid="a4086b1d-a96a-5d6b-8e4f-2030e6f25ba6"))
+    Dependency("Lua_jll")
+    Dependency(PackageSpec(name="oneTBB_jll", uuid="1317d2d5-d96f-522e-a858-c73665f53c3e"))
+    Dependency(PackageSpec(name="Expat_jll", uuid="2e619515-83b5-522b-bb60-26c02a35a201"))
```
Comment (giordano) [inline O/OSRM/build_tarballs.jl:53]: #3024

**Comment by mattwigway** [comment]
@jeremiahpslewis that's great! Yes, I need the libs as well. Let me know if there's anything I can do to help. I've been working on creating a JLL myself that's nearly done, perhaps we can combine efforts.

**Comment by jeremiahpslewis** [comment]
> @jeremiahpslewis that's great! Yes, I need the libs as well. Let me know if there's anything I can do to help. I've been working on creating a JLL myself that's nearly done, perhaps we can combine efforts.

Gladly, perhaps this PR is a good starting point for a collaboration? Do you have any idea how to get the osrm build to output dynamically linked libraries?

**Comment by jeremiahpslewis** [comment]
@mattwigway Just gave you access to the branch / fork so you can make changes.

**Comment by mattwigway** [comment]
@jeremiahpslewis thanks, I'll take a look. Allegedly `-DBUILD_SHARED_LIBS=ON` should do the trick, but I'm getting some linker errors when I try that.

**Comment by mattwigway** [comment]
I also found that using the toolchain file for CMake avoided me needing to specify all the library locations: https://github.com/JuliaPackaging/BinaryBuilder.jl/blob/master/docs/src/build_tips.md#cmake-builds

**Comment by mattwigway** [comment]
Attempting a build with the shared libraries now.

**Comment by jeremiahpslewis** [comment]
I just ran it locally with your -D command, seems to work, so I've added the library products as well. :)

**Comment by mattwigway** [comment]
It's possible the error I was running into is apple M1 specific - will see momentarily

**Comment by mattwigway** [comment]
I was thinking it might also make sense to add the profiles (car.lua and friends) as FileProducts since OSRM is not very useful without these.

**Comment by jeremiahpslewis** [comment]
> I was thinking it might also make sense to add the profiles (car.lua and friends) as FileProducts since OSRM is not very useful without these.

Sure, fine by me. Want to add it to the PR yourself?

**Comment by mattwigway** [comment]
Will do.

**Comment by mattwigway** [comment]
I'm building on x86, but the aarch64 mac target is failing. I think this is an upstream issue though. I'm trying to build OSRM outside of BinaryBuilder on an actual M1 Mac, and if that fails too I'll file an issue with the OSRM project. I think if that's the case we should just remove aarch64-apple-darwin as a supported platform from the jll for now.

**Comment by mattwigway** [comment]
Yep, same error when building natively on M1.

**Comment by jeremiahpslewis** [comment]
@mattwigway ok, so we are ✅ for linux, but windows and Mac seem to be blocked (upstream?)

Release plan:
- Get a new OSRM release tagged https://github.com/Project-OSRM/osrm-backend/issues/6516
- Merge this PR and ship binary for linux
- Once mac / windows builds are debugged, then add and ship them

**Comment by mattwigway** [comment]
Thanks! I'll look quickly at the Windows build situation as it would be
ideal if I had a Windows binary available (I know, I know, I hate Windows
too... but unfortunately for the project I'm working on my hands are tied).

**Comment by mattwigway** [comment]
No luck getting the Windows binaries built - endless errors with mingw redefining GCC internal functions, and TBB not finding things from `float.h`. I'm giving up on Windows for now.

**Comment by jeremiahpslewis** [comment]
@mattwigway the latest oneTBB is also a disaster for windows, so until someone upstream makes this happen, think we‘ll have to stay content with Linux (and Mac someday?)

do you know anyone you could ping to get a new version tagged of osrm?

**Comment by giordano** [comment]
> and TBB not finding things from float.h

https://github.com/JuliaPackaging/Yggdrasil/pull/6115#issuecomment-1399079196

**Comment by jeremiahpslewis** [comment]
Mac (aarch64 error):

```bash
[01:10:34]   "_crc32_combine", referenced from:
--
  | [01:10:34]       __ZN3tbb6detail2d115concrete_filterINSt3__110shared_ptrIZN4osrm8guidance13annotateTurnsERKNS5_4util12DynamicGraphINS7_17NodeBasedEdgeDataEEERKNS5_9extractor6detail30EdgeBasedNodeDataContainerImplILNS5_7storage9OwnershipE0EEERKNS3_6vectorINS7_10CoordinateENS3_9allocatorISM_EEEERKNSD_23CompressedEdgeContainerERKNS3_13unordered_setIjNS3_4hashIjEENS3_8equal_toIjEENSN_IjEEEERKNSD_18NodeRestrictionMapINSD_17UnconditionalOnlyEEERKNSD_17WayRestrictionMapERKNSE_13NameTableImplILSH_0EEERKNSD_11SuffixTableERKNS3_5tupleIJNSL_IjS10_EENSL_ItNSN_ItEEEEEEERNS7_15ConcurrentIDMapIS1M_tNSD_24TurnLaneDescription_hashEEERNS1Q_INS7_8guidance15LaneTupleIdPairEtN5boost4hashIS1V_EEEERNS6_6detail21TurnDataContainerImplILSH_2EEERS1K_RNS1Q_INS1U_12BearingClassEjNSW_IS26_EEEERNS1Q_INS1U_10EntryClassEtNSW_IS2A_EEEERjE19TurnsPipelineBufferEEvZNS6_13annotateTurnsESC_SK_SR_SU_S13_S18_S1B_S1F_S1I_S1P_S1T_S20_S24_S25_S29_S2D_S2E_E3$_2EclEPv in guidance_processing.cpp.o
  | [01:10:34] ld: symbol(s) not found for architecture arm64
```

This seems to be a `zlib` linking issue, but nothing I've tried so far resolves it.

**Comment by giordano** [comment]
https://github.com/JuliaPackaging/Yggdrasil/pull/5425#issuecomment-1242994320

> If you want to make everybody's life easier, when you show linking errors you also want to show the compiler invocation line.  Because the `undefined reference to` error means that the linker can't find the mentioned symbol in any of the linked libraries, which often simply means that the library which is supposed to implement the missing symbol wasn't passed as argument.  But if I don't see the compiler invocation I can't tell you "see, that's the problem".

**Comment by jeremiahpslewis** [comment]
> [#5425 (comment)](https://github.com/JuliaPackaging/Yggdrasil/pull/5425#issuecomment-1242994320)
> 
> > If you want to make everybody's life easier, when you show linking errors you also want to show the compiler invocation line.  Because the `undefined reference to` error means that the linker can't find the mentioned symbol in any of the linked libraries, which often simply means that the library which is supposed to implement the missing symbol wasn't passed as argument.  But if I don't see the compiler invocation I can't tell you "see, that's the problem".

Thanks, will update!

**Comment by jeremiahpslewis** [comment]
Mingw error, which also pops up on `apple-darwin`; adding `-fext-numeric-literals` to `CMAKE_CXX_FLAGS` didn't seem to help.
```log
/opt/x86_64-w64-mingw32/x86_64-w64-mingw32/sys-root/usr/local/include/boost/math/constants/constants.hpp:246:3: error: unable to find numeric literal operator ‘operator""Q’[39m

  246 |   BOOST_DEFINE_MATH_CONSTANT(half, 5.000000000000000000000000000000000000e-01, "5.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e-01")[39m

      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~[39m

/opt/x86_64-w64-mingw32/x86_64-w64-mingw32/sys-root/usr/local/include/boost/math/constants/constants.hpp:246:3: note: use ‘-fext-numeric-literals’ to enable more built-in suffixes
```

**Pull request closed without merging**